L74-java-channel-creds.md

```diff
@@ -0,0 +1,478 @@
+channel/server construction.
+
+This is to allow:
+
+*   implementing new transport credentials without creating a new
+    channel/server type, which allows users of the API to separate concerns
+*   providing stable APIs for advanced credential configuration like mutual
+    TLS, without polluting the general-purpose ManagedChannelBuilder API with
+    many overloads that don't apply to all transports
+*   composition of credentials, like necessary for xDS where a fallback
+    credential needs to be provided by the user
+*   aligning with C core and Go implementations
+
+_The design focuses on channel credentials, since they are the most complex and
+most discussed._ Server credentials will mostly just mirror the channel
+credentials; they suffer from most of the same problems but at a smaller
+scale.
+
+## Background
+
+Prior to gRPC v0.12.0, C core had a "unified" `Credential` type that was
+simultaneously a call and channel credential. This was the design proposed for
+all languages. However, the Java implementation could not determine how such an
+API could be implemented for Java with its multiple transports. The designer of
+the API had too little bandwidth (it was a _very_ busy time), so Java (and Go)
+developed its own API. Java used interceptors (which no other language had) and
+credential-specific configuration on the ManagedChannelBuilder.
+
+In gRPC v0.12.0, C core swapped to a “split” CallCredential and
+ChannelCredential. ChannelCredentials could contain CallCredentials. At this
+point it was not a priority for Java to investigate the new API; Java was happy
+they didn’t use the previous API, such that they didn’t have an API migration.
+
+gRPC Java 0.15.0 added CallCredentials to replace the interceptor-based call
+credentials. At this point Java had looked into the full “split” credential
+design, but decided it was too hard to implement at that moment. Java worked to
+stabilize the `ManagedChannelBuilder.usePlaintext()` and
+`useTransportSecurity()` APIs.
+
+Go has split call credentials (`credentials.PerRPCCredentials`) and channel
+credentials (`credentials.TransportCredentials`), but with a separate type to
+combine them (`credentials.Bundle`). The split credentials significantly
+predate `Bundle`.
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+`ChannelCredentials` will be introduced, which can be provided when
+constructing a `ManagedChannelBuilder':
+
+```java
+// instead of:
+channel = ManagedChannelBuilder.forTarget("example.com")
+    .build();
+channel = ManagedChannelBuilder.forTarget("example.com").usePlaintext()
+    .build();
+channel = ManagedChannelBuilder.forAddress("example.com", 443)
+    .build();
+// the user would now:
+channel = Grpc.newChannelBuilder("example.com", new TlsChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilder("example.com", new InsecureChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilderForAddress("example.com", 443, new TlsChannelCredentials())
```

Comment:
In the future, are you expecting the server builder to be constructed in the same manner?  You mentioned there were channel cred similarities above, which is why I ask

Reply:
I actually have some snippets with what it will look like at the bottom `Grpc.newServerBuilderForPort()`. I didn't provide reasoning, but I purposefully did not use `Grpc.newServerBuilder()` as I want to keep that open for future expansion (I still have the long-term goal of allowing adding/removing listeners on a started server, in which case no port would be required when building the server).

TlsServerCredentials were annoying to figure out how I wanted them to look. We still have the "Java can't easily parse PKCS#8", so I wanted to keep private key parsing code in the transports and not in grpc-api.